### PR TITLE
[Feature #72] Add dockerization of fta-fmea-ui

### PR DESCRIPTION
**Pull request by Matthew-Kulich**

### CODE REVIEW

File: deploy/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+COPY package-lock.json ./
+
+ENV SERVER_URL="http://localhost:9999"
+ENV CONTEXT="/fta-fmea"
+ENV ADMIN_ONLY=false
+
+RUN npm install
+COPY ./ ./
+
+RUN REACT_APP_BASE_API_URL=${SERVER_URL}/${CONTEXT}-server \
```
Comment (blcham) [inline deploy/Dockerfile:14]: OK definitely very good step forward,  but this is still creating build-time dependencies and not runtime dependencies as we need.

I think you will understand as you will be creating `docker-compose.yml` file. I assume you did not do that, see A/C of the ticket:
> there is docker-compose file in this repository which orchestrates backend and frontend together

I suggest to try to create docker-compose.yml first and when done also tag @LaChope to review the pull request, then if needed agree on some short meeting to explain each other ideas.

Comment (blcham) [reply deploy/Dockerfile:14]: Of course, i am also available to discuss ...

**Review by blcham — CHANGES_REQUESTED** [review]
Looks already quite good, but not finished yed due to missing docker-compose.yml

File: deploy/Dockerfile (hunk 1 above)
Comment (Matthew-Kulich) [reply deploy/Dockerfile:14]: No, i didnt do that. 
Ok, thanks.

**Comment by Matthew-Kulich** [comment]
Look at the last commit and see if it is ok, please. I guess these iterations will be the best for me to learn this.
@blcham @LaChope

File: deploy/Dockerfile [hunk 2]
```diff
@@ -1,26 +1,27 @@
 FROM node:16-alpine AS BASE
 WORKDIR /usr/src/app
-
-COPY ../package.json ./
-COPY ../package-lock.json ./
+COPY package.json package-lock.json ./
```
Comment (blcham) [inline deploy/Dockerfile:3]: The existence of Dockerfile in this directory is at least very unusual here. 

To me it make sense to have such Dockerfile in root project of this folder. On push to master we should trigger this `docker build` which should deploy image to the docker github registry.  Docker-compose within `./deploy` folder should use the image from docker registry directly.

The setting described above you can use without docker registry as follows:
```
cd $project_root
docker build .
cd ./deploy
docker-compose up
```

File: README.md [hunk 3]
```diff
@@ -36,3 +36,10 @@ There are two flags which need to be set:
 ## Related links
 * Running demo application - [kbss/fta-fmea-demo](https://kbss.felk.cvut.cz/fta-fmea-demo/) 
 * Backend project - [fta-fmea](https://github.com/kbss-cvut/fta-fmea)
+
+## Dockerization
+1. Download [GraphDB](https://graphdb.ontotext.com/) standalone server ZIP archive and place it into the `deploy/db-server` folder.
```
Comment (blcham) [inline README.md:41]: Just curious if we can have a shell script called 
'./deploy/db-server/bin/download-graphdb.sh` script to do that from environment settings ?

File: deploy/nginx/Dockerfile [hunk 4]
```diff
@@ -1,6 +0,0 @@
-FROM nginx:latest
```
Comment (blcham) [inline deploy/nginx/Dockerfile]: Not sure why removing this. Hope I did not confuse u to do that :) We can discuss on standup.

**Review by blcham — CHANGES_REQUESTED** [review]
I see multiple improvements, but still there are some design issues here and it seems right time to discuss it.

File: README.md (hunk 3 above)
Comment (Matthew-Kulich) [reply README.md:41]: Yes, we can, ill do that

File: deploy/Dockerfile [hunk 5]
```diff
@@ -25,3 +25,7 @@ COPY --from=build /usr/src/app/public/index.html /usr/share/nginx/html/index.htm
 COPY --from=build /usr/src/app/public/build/bundle.js /usr/share/nginx/html/build/bundle.js
 
 RUN chmod a+r -R /usr/share/nginx/html
+
+COPY .docker/docker-entrypoint.sh /
+ENTRYPOINT ["/docker-entrypoint.sh"]
```
Comment (blcham) [inline deploy/Dockerfile:30]: @Matthew-Kulich @LaChope It was not possible to use nginx envsubst enpoint?

Comment (LaChope) [reply deploy/Dockerfile:30]: @blcham It is not working yet, we are just still under investigation :)

Comment (blcham) [reply deploy/Dockerfile:30]: oh sry ;)

File: deploy/Dockerfile [hunk 6]
```diff
@@ -1,26 +1,27 @@
...
-ENV CONTEXT="/fta-fmea"
+ENV CONTEXT="fta-fmea"
 ENV ADMIN_ONLY=false
 
 FROM base AS dependencies
 
 RUN npm install
 COPY ./ ./
 
+RUN sed -i "s|BASE_API_URL=.*$|BASE_API_URL=$SERVER_URL/$CONTEXT|" .env.prod
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply deploy/Dockerfile]: Implemented the "ugly" way, later we can make it prettier

File: README.md [hunk 7]
```diff
@@ -36,3 +36,10 @@ There are two flags which need to be set:
...
+2. Set `GRAPHDB_ZIP` variable in `deploy/db-server/Dockerfile` to the name of the file you downloaded.
+3. Clone [fta-fmea](https://github.com/kbss-cvut/fta-fmea) as a subfolder of the 'deploy' folder, i.e.:
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply README.md:43]: Done

File: README.md [hunk 8]
```diff
@@ -36,3 +36,10 @@ There are two flags which need to be set:
 ## Related links
 * Running demo application - [kbss/fta-fmea-demo](https://kbss.felk.cvut.cz/fta-fmea-demo/) 
 * Backend project - [fta-fmea](https://github.com/kbss-cvut/fta-fmea)
+
+## Dockerization
+1. Download [GraphDB](https://graphdb.ontotext.com/) standalone server ZIP archive and place it into the `deploy/db-server` folder.
+2. Set `GRAPHDB_ZIP` variable in `deploy/db-server/Dockerfile` to the name of the file you downloaded.
```
[the remark this answers is not in the document]
Comment (Matthew-Kulich) [reply README.md:42]: Done

File: README.md (hunk 3 above)
Comment (Matthew-Kulich) [reply README.md:41]: We can't download it directly, first we need to fill form and then download it from email.
And I'm not sure how long this link will be available.

The link from email looks like : 
https://email.ontotext.com/e3t/Ctc/GD+113/cGJhF04/VWcV8x3WHrJ3W8mhwjv35SMRCW8cYYgJ4L1WwfN39GZDG3q3pBV1-WJV7CgFvbW1WVgFZ4YHJL6N3g9W85tCDKRN2VXCjph3CTSVf-kjf8vBTNKW6MdSGT8-7vy0W8bsPBX84NDWJW3jJcyL7KcTYFW1ghc-p5Ttt8zW1fvmTN1lNjscW8ywD756g74N9W3pQ8GG6-JQwYW1H5cwQ6FvvKkW35KyNK69t1kjW974Bjy98rB8HW62KhtP1wZtT-W5f6d2B5t_6GcW6kp3DR1z-fjXW18xKrH8LPS27W2q3NcY5WrQnDW2D_n4Y4dPVWDW2YzlbK8hJYBbW4wYkMj6BDGzkW5xGyL57vq8-dN6gBw_ppXD9YW36jWGX8XNjbVW5GGBHx11m8KqW8NqzzL7p3YLfW7xmPNp4YsZWrW186bFM9b12ZGW3-4-Ks1tgcQ83cBj1

@blcham

**Review by blcham — APPROVED** [review]
LGTM

**Pull request merged**